tbtadm/controller.cpp

```diff
@@ -653,6 +655,188 @@ catch (...)
     m_err << "Unknown exception\n";
 }
 
+void tbtadm::Controller::addToBootACL(const fs::path& dir, std::string uuid)
+{
+    auto fname = dir.filename().string();
+    auto num = fname.substr(0, fname.find('-'));
+    auto boot_acl_path = sysfsDevicesPath / (domain + num) / boot_acl;
+
+    if (!fs::exists(boot_acl_path))
+    {
+        /* Cannot find boot_acl */
+        return;
+    }
+
+    /* Read boot_acl */
+    std::stringstream stream(readAndTrim(boot_acl_path));
+    /* Check maximum allowed boot_acl string size */
+    if (stream.str().size() > 36 * 16 + 1 * 15)
+    {
+        m_out << "boot_acl contains too much characters\n";
+        return;
+    }
+
+    std::queue<std::string> acl_queue;
+
+    while(stream.good())
+    {
+        std::string substring;
+
+        getline(stream, substring, ',');
+	/* this moves also uuid to the end of the queue */
+        if (substring.size() == 36 && substring != uuid)
+        {
+            acl_queue.push(substring);
+        }
+    }
+
+    if (acl_queue.size() > 16)
+    {
+        m_out << "Too much elements in the boot_acl list\n";
+        return;
+    }
+
+    /* Add UUID to boot_acl queue */
+    if (acl_queue.size() < 16)
+    {
+        acl_queue.push(uuid);
+    }
+    else
+    {
+        m_out << "Trim boot_acl list, remove oldest entry\n";
+        acl_queue.pop();
+        acl_queue.push(uuid);
+    }
+
+    std::string out;
+
+    for (int i = 0; i < 16; i++)
+    {
+        if (!acl_queue.empty())
+        {
+            out.append(acl_queue.front());
+            acl_queue.pop();
+        }
+
+        if (i < 15)
+        {
+            out.append(",");
+        }
+    }
+
+    std::ofstream boot_acl_file(boot_acl_path.string());
+    boot_acl_file << out;
+    boot_acl_file.close();
+}
+
+std::string tbtadm::Controller::getDomain(std::string uuid)
+{
+    if (!fs::exists(sysfsDevicesPath))
+    {
+        return "";
+    }
+
+    for (auto dir : boost::make_iterator_range(
+                fs::directory_iterator(sysfsDevicesPath), {}))
+    {
+        if (dir.status().type() != fs::directory_file)
+        {
+            continue;
+        }
+
+	if (!isDevice(dir.path()))
+	{
+		continue;
+	}
+
+        auto routeString = dir.path().filename().string();
+        auto domain_num = routeString.substr(0, routeString.find('-'));
+
+        if (readAndTrim(dir / uniqueIDFilename) == uuid)
+        {
+            /* Consider operate with integer
+             * return std::stoi(domain_num, nullptr);
+             */
+            return domain_num;
+        }
+    }
+
+    return "";
+}
+
+void tbtadm::Controller::removeFromBootACL(std::string uuid)
+{
+    /* Get domain associated with uuid */
```

Comment:
Nope, it should be removed from all the domains it appears in

Reply:
@westeri btw do we need entries for every domain/controller if they are effectively should be identical?

Reply:
Yes, because it's possible to have one controller with boot ACL support and another one without it (depending on the FW version, for example) or to have different number of entries supported.

From driver's perspective, it also gives the flexibility for the user to have different entries in each one, if the user wants to build a complex setup with many devices so there is not enough place for them in the same boot ACL list, so the user can choose to add some devices to this controller and some others to the other one. Mechanism, not policy, as we know.

For user-space, it's fine that we start with "keep all the controllers in sync". Later, we may decide to add some knobs and switches for fine-tuning it, if users want such a functionality.